synapse/rest/media/v1/preview_html.py

```diff
@@ -323,10 +323,13 @@ def _iterate_over_text(
 def rebase_url(url: str, base: str) -> str:
     base_parts = list(urlparse.urlparse(base))
     url_parts = list(urlparse.urlparse(url))
-    if not url_parts[0]:  # fix up schema
+    # Add a scheme, if one does not exist.
+    if not url_parts[0]:
         url_parts[0] = base_parts[0] or "http"
-    if not url_parts[1]:  # fix up hostname
+    # Fix up the hostname, if this is not a data URL.
+    if url_parts[0] != "data" and not url_parts[1]:
         url_parts[1] = base_parts[1]
+        # If the path does not start with a /, nest it under the base path.
         if not url_parts[2].startswith("/"):
             url_parts[2] = re.sub(r"/[^/]+$", "/", base_parts[2]) + url_parts[2]
```

Comment:
`url_parts` seems to be a named tuple these days. I don't knowif you'd consider `url_parts.schema` an improvement over `url_parts[0]` though.

C.f. [docs](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) and [typeshed](https://github.com/python/typeshed/blob/fe6233a8c3d4ca708a0da2ab281a20fbd148215c/stdlib/urllib/parse.pyi#L56-L62)

Reply:
I attempted to change this code, but we actually modify the pieces, which is why it is a list. 🤷 Not converting `base_parts` to a list and using the named properties might be clearer though?

Reply:
I made the change to `base_parts`.